### PR TITLE
[1.5] Add support for XPath by introducing `Selector` and `Mouse::findElement()`

### LINKED ISSUE #333: Xpath support
Hi, what would you say to adding a support for xpath to the current css selectors?

We use js already to find elements using css selectors.

Looking for elements using xpath seems to be as easy as

```js
var xpath = "//a[text()='SearchingText']";
var matchingElement = document.evaluate(xpath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
```

I propose to add `Selector` interface with `CssSelector` and `XPathSelector` implementations, used as

`Mouse::find(Selector $selectors, int $position = 1): self`

in v2.0

### PR DESCRIPTION
**Pull request by simPod**
Resolves https://github.com/chrome-php/chrome/issues/333

### CODE REVIEW

File: src/Dom/Selector/Selector.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace HeadlessChromium\Dom\Selector;
+
+interface Selector
+{
+    public function expressionCount(): string;
```
Comment (enricodias) [inline src/Dom/Selector/Selector.php:9]: Why use string here?

File: src/Input/Mouse.php [hunk 2]
```diff
@@ -260,26 +262,54 @@ private function scrollToBoundary(int $right, int $bottom): self
...
+            $position = \max(1, $position);
+            $position = \min($position, $elementCount);
 
             $element = $this->page
-                ->evaluate('JSON.parse(JSON.stringify(document.querySelectorAll("'.$selectors.'")['.$position.'].getBoundingClientRect()));')
+                ->evaluate(\sprintf('JSON.parse(JSON.stringify(%s.getBoundingClientRect()));', $selector->expressionFindOne($position)))
                 ->getReturnValue();
         } catch (JavascriptException $exception) {
-            throw new ElementNotFoundException('The search for "'.$selectors.'" returned no elements.');
+            throw new ElementNotFoundException('The search for "'.$selector->expressionCount().'" returned no result.');
```
Comment (enricodias) [inline src/Input/Mouse.php:299]: The message for this exception will not mention the selector name, will it?

**Comment by enricodias** [comment]
Do you think it's worth it to have a test class for each `Selector` class separately? Even thought they are still being tested indirectly in the mouse tests.

File: src/Dom/Selector/Selector.php (hunk 1 above)
Comment (simPod) [reply src/Dom/Selector/Selector.php:9]: it returns an expression

File: src/Input/Mouse.php (hunk 2 above)
Comment (simPod) [reply src/Input/Mouse.php:299]: https://github.com/chrome-php/chrome/pull/340#discussion_r795632223

**Comment by simPod** [comment]
> Do you think it's worth it to have a test class for each Selector class separately? Even thought they are still being tested indirectly in the mouse tests.

It discovered few bugs related to position when I deved it.

**Comment by simPod** [comment]
Do we want to pursue this?

**Comment by enricodias** [comment]
I think so.

**Review by enricodias — APPROVED** [review]

**Comment by simPod** [comment]
I have rebased on 1.5 since there was a force push I guess

**Comment by simPod** [comment]
_The scrolling tests in GA are flaky (see https://github.com/simPod/chrome/runs/5407170225)_

**Comment by enricodias** [comment]
Yes, they are. But I never managed to reproduce this error locally :/

**Review by enricodias — APPROVED** [review]

**Comment by simPod** [comment]
> Yes, they are. But I never managed to reproduce this error locally :/

yup, same here

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
Are you able to resolve the merge conflicts, please?

**Pull request merged**